flareon/src/router/path.rs

```diff
@@ -15,30 +14,38 @@ impl PathMatcher {
     pub(crate) fn new<T: Into<String>>(path_pattern: T) -> Self {
         let path_pattern = path_pattern.into();
 
-        let mut last_end = 0;
+        #[derive(Debug, Copy, Clone)]
+        enum State {
+            Literal { start: usize },
+            Param { start: usize },
+        }
+
         let mut parts = Vec::new();
-        let param_regex = Regex::new(":([^/]+)").expect("Invalid regex");
-        for capture in param_regex.captures_iter(&path_pattern) {
-            let full_match = capture.get(0).expect("Could not get regex match");
-            let start = full_match.start();
-            if start > last_end {
-                parts.push(PathPart::Literal(path_pattern[last_end..start].to_string()));
+        let mut state = State::Literal { start: 0 };
+
+        for (index, ch) in path_pattern.chars().map(Some).chain([None]).enumerate() {
+            match (ch, state) {
+                (Some('/') | None, State::Param { start }) => {
+                    let param_name = &path_pattern[start..index];
+                    assert!(
+                        Self::is_param_name_valid(param_name),
+                        "Invalid parameter name: `{param_name}`"
+                    );
+
+                    parts.push(PathPart::Param {
+                        name: param_name.to_string(),
+                    });
+                    state = State::Literal { start: index };
+                }
+                (Some(':') | None, State::Literal { start }) => {
+                    let literal = &path_pattern[start..index];
+                    if !literal.is_empty() {
+                        parts.push(PathPart::Literal(literal.to_string()));
+                    }
+                    state = State::Param { start: index + 1 };
+                }
+                _ => {}
```

Comment:
Imho, it would be good to cover some known incorrect states and err on them, eq. `(Some(':') | None, State::Param { start })` which would happen if someone did `/todos/:todo_id:/remove`

Reply:
This particular case is already covered by `assert!(Self::is_param_name_valid(param_name))` in line 31.

Reply:
Ah, sorry, I just got your comment — while `is_param_name_valid` will indeed fail on this, it would be nice to get a better error message. I'll fix this.

Reply:
This is done now in the latest revision.